observer/lib/observer.js

```diff
@@ -0,0 +1,39 @@
+import { updateDailyTransferStats } from './platform-stats.js'
+
+/**
+ * Observe the transfer events on the Filecoin blockchain
+ * @param {import('pg').Pool} pgPool
+ * @param {import('ethers').Contract} ieContract
+ * @param {import('ethers').Provider} provider
+ */
+export const observeTransferEvents = async (pgPool, ieContract, provider) => {
+  const { rows } = await pgPool.query(
+    'SELECT MAX(last_checked_block) FROM daily_reward_transfers'
+  )
+  const lastCheckedBlock = rows[0].last_checked_block
+
+  console.log('Querying impact evaluator Transfer events after block', lastCheckedBlock)
+  let events
+  try {
+    events = await ieContract.queryFilter(ieContract.filters.Transfer(), lastCheckedBlock)
+  } catch (error) {
+    console.error('Error querying impact evaluator Transfer events', error)
+    if (error.message.includes('bad tipset height')) {
+      console.log('Block number too old, GLIF only provides last 2000 blocks, querying from -1900')
+      events = await ieContract.queryFilter(ieContract.filters.Transfer(), -1900)
+    } else {
+      throw error
+    }
+  }
+  const currentBlockNumber = await provider.getBlockNumber()
+  console.log('Current block number:', currentBlockNumber)
+  console.log(`Found ${events.length} Transfer events`)
+  for (const event of events) {
+    const transferEvent = {
+      to_address: event.args.to,
+      amount: event.args.amount
+    }
+    console.log('Transfer event:', transferEvent)
+    await updateDailyTransferStats(pgPool, transferEvent, currentBlockNumber)
```

Comment:
Making one SQL query per participant is expensive. We have ~5k daily participants. Not all of them will receive rewards every week, but still...

We need to check how long does it take to run all these SQL queries.

If it takes too long, then we should rework `observeTransferEvents` to batch SQL queries.

Ideally, we should wait until we collect all events emitted in a single block number, and then run a single SQL query to update all participants that received rewards in that block.

@juliangruber @PatrickNercessian

Reply:
+1 to batching, after measuring the impact. Can you run a benchmark with 100k participants and batching vs no batching?

Reply:
It took ~24.8 seconds to update the table 100,000 times. Tried again with 1,000 updates, and it was 1.4 seconds.

This is of course on my machine, so there may be differences depending on hosting environment. But it seems pretty negligible?

I am open to making the batching change if we prefer it for code quality purposes, even if it doesn't have significant performance gains.

Here was my benchmarking code, if it's of interest:

```
import { updateDailyTransferStats } from '../lib/platform-stats.js'
import { getPgPool } from '../lib/db.js'

const pgPool = await getPgPool()

await pgPool.query('DELETE FROM daily_reward_transfers')

let timeToUpdateTable = 0
for (let index = 0; index < 5_000; index++) {
  const transferEvent = {
    toAddress: "This is fake mock address " + index,
    amount: Math.floor(Math.random() * Number.MAX_SAFE_INTEGER)
  }
  console.log('Transfer event index:', index, 'Transfer event:', transferEvent)
  const beforeTime = Date.now()
  await updateDailyTransferStats(pgPool, transferEvent, 3988923)
  timeToUpdateTable += Date.now() - beforeTime
}

console.log('Time to update table:', timeToUpdateTable)

```
await pgPool.end()

Reply:
Ok thank you for testing this, I believe this is good enough

Reply:
Thank you for measuring the duration, @PatrickNercessian!

I agree with you that 1.4 seconds to update 5000 participants is acceptable.

Could you please ensure we can tell how long this is taking in production? We can keep it super simple - we already log a message before calling the SQL, so let's add another console log after the SQL query is done. Our log service will add timestamps to these messages, and then we can inspect the logs to understand how long it took to observe a weekly payout.

Reply:
We released the rewards earlier today, it seems the entire process of observing the smart contract events finished under one second for ~300 participants. All is good, no code changes are necessary.